### ISSUE TITLE
Stuck in threading on macOS

### ISSUE THREAD

**Issue by kpj** [jaseg/python-mpv#61]
Hello,

consider the following code:
```python
import mpv

_url = 'test.webm'
mpv = mpv.MPV()
mpv.loadfile(_url)
print('waiting')
mpv.wait_for_playback()
print('done')
```

On Linux, this will print `waiting`, then play `test.webm`, and finally print `done` as expected.

On macOS 10.12.6 using Python 3.6 and MPV 0.27 however, it only prints `waiting` and is then stuck forever while the Python-Rocket starts jumping in my Dock.

I looked into python-mpv's code, and for `mpv.wait_for_playback()` it is stuck in [line 557](https://github.com/jaseg/python-mpv/blob/master/mpv.py#L557), which is:
```python
self._playback_cond.wait()  # _playback_cond is threading.Condition
```
Using `mpv.wait_for_property('filename', lambda x: x is None)` instead, makes it stuck in [line 569](https://github.com/jaseg/python-mpv/blob/master/mpv.py#L569), which is:
```python
sema.acquire()  # sema is threading.Semaphore
```

The problem thus seems to somehow be related to the handling of the `threading` library.
Do you have any suggestion what might be causing this?

This is also possibly a duplicate of #59.

**Comment by jaseg** [comment]
The semaphore is most likely working alright. That is part of python and that is probably well tested even on osx. The problem is most likely in the [event handling](https://github.com/jaseg/python-mpv/blob/master/mpv.py#L422). I'm a bit challenged here since I've never used osx and I don't have any osx machines available. Could you try running your script in gdb to get a stack trace of the event handler thread? See [this comment](https://github.com/jaseg/python-mpv/issues/36#issuecomment-314681653) for instructions.

**Comment by jaseg** [comment]
The bouncing launcher you mentioned might be evidence that libmpv tried to create a window, but for some reason did not succeed. This sounds vaguely similar to what @Shu-ji describes in [this comment](https://github.com/jaseg/python-mpv/issues/36#issuecomment-346938221). Could you try the workaround described there (manually creating a PyQT window and passing the window ID to mpv)?

**Comment by kpj** [comment]
Thanks for your suggestions, I agree that a bug in Python's `threading` is highly unlikely.

### Event Handler Stack Trace
* I am using mpv 0.27 and libmpv (1, 25).
* Adding `mpv['vo'] = 'opengl'` to my previous example does not alter its behavior.
* I can currently not run GDB due to weird macOS-related problems (https://github.com/Homebrew/homebrew-core/issues/20047). I will try again, when I have more time.

### Manual PyQT window
Is there a way of doing this with your python-mpv version?

**Comment by jaseg** [comment]
You can use the linked example almost unmodified:
```python
#!/usr/bin/env python3
import mpv
import time
import os
import sys

from PyQt5.QtWidgets import *
from PyQt5.QtCore import *

class Test(QMainWindow):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.container = QWidget(self)
        self.setCentralWidget(self.container)
        self.container.setAttribute(Qt.WA_DontCreateNativeAncestors)
        self.container.setAttribute(Qt.WA_NativeWindow)
        player = mpv.MPV(wid=str(int(self.container.winId())),
                vo='x11',
                log_handler=print,
                loglevel='debug')
        player.play('test.webm')

app = QApplication(sys.argv)

import locale
locale.setlocale(locale.LC_NUMERIC, 'C')
win = Test()
win.show()

sys.exit(app.exec_())
```
You may not need the ```vo='x11'```. The setlocale is necessary since pyqt apparently stomps all over that on load. This could also be solved by just import pyqt before mpv (which already includes that line).

**Comment by kpj** [comment]
Neat, your example works nicely and plays the video on macOS for me (after removing `vo='x11',`).

How is the `wid` argument handled in python-mpv? It seems to be part of `extra_mpv_opts`, which is then somehow processed in [_mpv_set_option_string](https://github.com/jaseg/python-mpv/blob/master/mpv.py#L550).

Do you think that python-mpv should take care of this `QWidget`-creation itself, in order to provide a fix for macOS?
Or one could try to debug this even further, using gdb.

**Comment by jaseg** [comment]
Yep, wid is effectively passed in as a command line option. See also [this upstream example](https://github.com/mpv-player/mpv-examples/blob/master/libmpv/cocoa/cocoabasic.m).

According to the README of [the upstream examples](https://github.com/mpv-player/mpv-examples/tree/master/libmpv) a caveat seems to be that this way of embedding is not very stable, so you might also try embedding via OpenGL instead. The callbacks are supported by python-mpv. [Upstream issue #556](https://github.com/mpv-player/mpv/issues/556) gives some detail on mpv GUI initialization on OSX, though it still doesn't explain the hang observed here.

I would like to avoid putting PyQT-specific code into python-mpv. Since getting that up and running is very simple (3loc) I'd keep it in the README examples for now. For now, I think two things to add there are a) an OpenGL callback example and b) a hint to OSX users to create their own window.

As for proper debugging, yes, that would be great. However, I don't have an Apple machine so you or one of the other OSX users would have to do most of that. And given what [upstream issue #556](https://github.com/mpv-player/mpv/issues/556) says we might well find out that maybe on OSX embedding is only supported if you bring your own window.

**Comment by kpj** [comment]
I agree, adding PyQT-specific code to handle macOS-specific problems does not seem sensible in this scenario.
For now I'll try to use this work-around instead of diving into more in-depth debugging craziness.

Assuming that I want to use python-mpv's `MPV` object in another application (for me that would be [here](https://github.com/kpj/Vydia/blob/425bbb566ae6d1c0ef72bb10a1ff8cd14d686d9c/vydia/extra/player.py#L141)), i.e. I don't want the PyQT event loop blocking my main thread.
What would be the best way of handling this case? At first, I thought about simply starting it in another thread (and then somehow try to access `MPV`'s `loadfile`, etc. methods):
```python
[..]

def foo():
    app = QApplication(sys.argv)

    import locale
    locale.setlocale(locale.LC_NUMERIC, 'C')
    win = Test()
    win.show()

    sys.exit(app.exec_())

threading.Thread(target=foo).start()
```

This however crashes:
```bash
2017-12-29 12:29:13.231 Python[21946:737607] *** Assertion failure in +[NSUndoManager _endTopLevelGroupings], /BuildRoot/Library/Caches/com.apple.xbs/Sources/Foundation/Foundation-1450.16/Foundation/Misc.subproj/NSUndoManager.m:361
2017-12-29 12:29:13.246 Python[21946:737607] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '+[NSUndoManager(NSInternal) _endTopLevelGroupings] is only safe to invoke on the main thread.'
```
with the main point probably being: `is only safe to invoke on the main thread`.

Do you have a suggestion for this, or is it again some macOS-specific difficulty?

**Comment by jaseg** [comment]
According to the [Qt doc](http://doc.qt.io/qt-5/thread-basics.html), Qt must run on the application's main thread. So this is not a limitation of python-mpv, mpv or even PyQt. I know this is inconvenient, but the best solution to this is probably to move other python work to auxiliary threads and possibly use PyQts Qt threading bindings.

A more low-effort alternative would be to farm out Qt code to a separate process using the [multiprocessing module](https://docs.python.org/3.6/library/multiprocessing.html). It sounds like that would be quite easy to implement in your use case.

**Comment by kpj** [comment]
I can indeed display the video using an extra `multiprocessing` process.
The problem then is however, that I still need to access the MPV object in the main process.

The simplest code (which abuses concurrency quite heavily) I could come up with to achieve this (in theory), is this:
```python
import sys
import multiprocessing

from PyQt5.QtWidgets import QWidget, QMainWindow, QApplication
from PyQt5.QtCore import Qt

import mpv


class Test(QMainWindow):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.container = QWidget(self)
        self.setCentralWidget(self.container)
        self.container.setAttribute(Qt.WA_DontCreateNativeAncestors)
        self.container.setAttribute(Qt.WA_NativeWindow)
        self.window_id = str(int(self.container.winId()))

def foo(player):
    app = QApplication(sys.argv)
    win = Test()
    win.show()

    mpv._mpv_set_option_string(player, 'wid', win.window_id)

    sys.exit(app.exec_())

player = mpv.MPV()
multiprocessing.Process(target=foo, args=(player,)).start()
player.play('test.webm')
```

Unfortunately, it crashes:
```bash
objc[61728]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called.
objc[61728]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

Playing around with arbitrary `time.sleep` calls or using `multiprocessing.Manager` has not helped me so far.
Do you have a suggestion?

**Comment by jaseg** [comment]
Yes. You cannot create the mpv object in the parent process, since it initializes the mpv handle. When using multiprocessing, import mpv and create the MPV object in the child process, then pass commands to the child process either low-level using [multiprocessing's pipes and queues](https://docs.python.org/3.6/library/multiprocessing.html#pipes-and-queues) or high-level using [multiprocessing's managers and proxies](https://docs.python.org/3.6/library/multiprocessing.html#proxy-objects).

**Comment by kpj** [comment]
Uuuff. This special treatment of macOS starts to become annoying :-P

Consider this thread-inside-a-process solution:

```python
def foo(pipe):
    # setup QT window
    app = QApplication(sys.argv)
    win = Test()
    win.show()

    # initialize MPV
    import mpv
    player = mpv.MPV()
    mpv._mpv_set_option_string(
        player.handle,
        'wid'.encode('utf-8'), win.window_id.encode('utf-8'))

    # poll pipe
    def handle_pipe():
        output_p, input_p = pipe
        while True:
            try:
                msg = output_p.recv()
                print(msg)
                if msg == 'play':
                    player.play('test.webm')
            except EOFError:
                break
    threading.Thread(target=handle_pipe).start()

    # run QT main-loop
    sys.exit(app.exec_())

# setup MPV
pipe = multiprocessing.Pipe()
multiprocessing.Process(target=foo, args=(pipe,)).start()

# communicate with MPV
output_p, input_p = pipe
time.sleep(2)
input_p.send('nonsense')
time.sleep(1)
input_p.send('play')
```
It waits 2 seconds, prints 'nonsense', waits another 2 seconds, prints 'play' and then plays the movie.

I had to use the thread inside of the process, in order to poll the pipe and run the QT main-loop at the same time.
This solutions seems very non-optimal to me. Would you have a suggestion how to improve it?

**Comment by jaseg** [comment]
Sorry for not actually improving your code, but I just fiddled around a bit and came up with the following:

**Comment by jaseg** [comment]
Sorry for not actually improving your code, but I just played around a bit and came up with the solution below. So far I tested it on Linux and it works fine there.

The main advantage of that is that it's rather simple and it provides access to most methods on an ```MPV``` instance by just adding them to the ```exposed``` array. The main disadvantages are that all the magic functions taking callables as well as direct property access don't work. Property access must be emulated with ```player._set_property('loop', 'inf')```. You could build your own ```multiprocessing.BaseProxy``` descendant improving that interface, but I don't think that'd be worth it.

```python
#!/usr/bin/env python3
import mpv
import time
import os
import sys

import multiprocessing
from multiprocessing.managers import BaseManager, BaseProxy

from PyQt5.QtWidgets import *
from PyQt5.QtCore import *

class MpvManager(BaseManager):
    pass

MpvManager.register('MPV', mpv.MPV,
    exposed=[
        'play',
        '_get_property',
        '_set_property'
    ])

class MpvWindowThing(QMainWindow):
    def __init__(self, manager, parent=None):
        super().__init__(parent)
        self.container = QWidget(self)
        self.setCentralWidget(self.container)
        self.container.setAttribute(Qt.WA_DontCreateNativeAncestors)
        self.container.setAttribute(Qt.WA_NativeWindow)
        print('Running as pid', os.getpid())
        player = manager.MPV(wid=str(int(self.container.winId())),
                vo='x11', # You may not need this
                log_handler=print,
                loglevel='debug')
        player._set_property('loop', 'inf')
        player.play('test.webm')

with MpvManager() as manager:
    app = QApplication(sys.argv)
    win = MpvWindowThing(manager)
    win.show()
    sys.exit(app.exec_())

sys.exit(0)
```

**Comment by kpj** [comment]
On macOS there seems to be a problem with `manager.MPV`.
It gets stuck on `v vo/opengl Initializing OpenGL backend 'cocoa'` and never plays the video.

If I replace it with `mpv.MPV` everything works as expected.

**Comment by jaseg** [comment]
Oh my. No, what happens is ```multiprocessing.manager.BaseManager``` creates a proxy thing for ```MPV``` such that when you do ```manager.MPV``` you get a proxy for a ```MPV``` object living inside a subprocess. This way you have the PyQT process running in the parent process and mpv running in the subprocess. I have no idea why that doesn't work on mac.

I'm sorry I don't have any better suggestions. All ways I could think of to replace your custom ```multiprocessing.Pipe``` logic with the autogenerated proxies from ```multirprocessing.managers.BaseManager``` are ugly as sin as ```multiprocessing``` really does not have a very good or flexible API. I'd definitely go for your way now that I've had a read through multiprocessing's source.

I guess your ```MPVProxy``` really is fine. I'd maybe factor out the receive loop into another class. 50loc for that sort of workaround is ok I guess.

**Comment by kpj** [comment]
I see. I now ended up with [this](https://github.com/kpj/Vydia/blob/master/vydia/extra/macos_mpv_wrapper.py) implementation.

It's not as flexible as I want it to be (e.g. attribute setting doesn't work properly yet, setting time-observers fails as lambdas cannot be pickled, ...), but it's a start.

**Comment by iwalton3** [comment]
If anyone is looking for a workaround to this, my [external MPV library](https://github.com/iwalton3/python-mpv-jsonipc) works on OSX. It implements a decent amount of the API of `python-mpv`, but it probably isn't a drop-in replacement. (It was originally implemented to allow MPV support on platforms where getting a build of libmpv1 is a pain.)

**Comment by jb-alvarado** [comment]
Hi, is there any new on this topic or an workaround? The PyQt solution is not really an option for me.

**Comment by roachcord3** [comment]
This issue has been bugging me so much that ~~I have opened a $250 bounty for anyone who solves this (in a way that ends up also resolving the issue for hydrusnetwork/hydrus#1132, i.e., gets an embedded player working in a `QWidget`.)~~ (The bounty site is no longer working. Frankly, it seems that it's attracting vibecoder vultures who use Claude to make their GitHub comments because they aren't capable of stringing together their own sentences, so it'd be irresponsible for me to try to offer the bounty anymore anyway. I will not honor this bounty anymore for any reason.)

FWIW, SMPlayer is a Qt based app that embeds MPV, apparently. I don't know what they do differently from python-mpv to make it work, but it's probably worth referencing their code: https://github.com/smplayer-dev/smplayer/blob/master/src/mplayerwindow.cpp

**Comment by thorlucas** [comment]
@roachcord3 I'm very glad you opened this bounty because I've been stuck with this issue for a while now.

**Comment by 0xa48rx394r83e9** [comment]
> This issue has been bugging me so much that I have opened a $250 bounty for anyone who solves this (in a way that ends up also resolving the issue for hydrusnetwork/hydrus#1132, i.e., gets an embedded player working in a `QWidget`.)
> 
> 
> 
> <a href="https://www.boss.dev/issue/github/MDU6SXNzdWUyODMyMjAyNzc="><img src="https://img.shields.io/endpoint.svg?url=https://api.boss.dev/badge/bounty%3Fgid%3Dboss%2Fhe1loz5bgrlekspz77w9i8" /></a>
> 
> 
> 
> FWIW, SMPlayer is a Qt based app that embeds MPV, apparently. I don't know what they do differently from python-mpv to make it work, but it's probably worth referencing their code: https://github.com/smplayer-dev/smplayer/blob/master/src/mplayerwindow.cpp

Hey, this is intriguing and something I'd be happy to look into, but first I'd need to know if a: this is still an issue (it's been two years) and b: if the bounty is still active.

**Comment by roachcord3** [comment]
> Hey, this is intriguing and something I'd be happy to look into, but first I'd need to know if a: this is still an issue (it's been two years) and b: if the bounty is still active.

Yes, it's still an issue. However, the bounty site doesn't work anymore, and I'm no longer offering the bounty (because it's attracting idiots using AI not only to write shit code, but also to communicate.)

**Comment by dude-k** [comment]
The original poster's example from Dec 2017 still doesn't work so, yes, this is still an issue. A fix that gets the example to work (or an example that's simpler than the QT approach) would be greatly appreciated. Thanks for looking into it. 

I could probably toss USD10 or something into a pot if you need a monetary motivation.

**Comment by jaseg** [comment]
From my end I can promise credit and thanks in the readme to whoever manages to solve this. I still don't have a mac, so I still can't work on this myself.

**Comment by betagan** [comment]
Leaving another working example here for reference. Not sure this is more helpful than the Qt example in general, but it might be useful if you're purpose-building for macOS.

```import mpv
from Cocoa import NSApplication

app = NSApplication.sharedApplication()

_url = 'test.webm'
mpv = mpv.MPV(log_handler=print, loglevel='info')
mpv.loadfile(_url)

@mpv.event_callback('shutdown')
@mpv.event_callback('end-file')
def stop_app(evt):
    app.terminate_(None)

app.run()
```


The VOs for mpv that are available (at least by default) for mpv on macOS are all cocoa based and require an existing instance of NSApplication (as https://github.com/mpv-player/mpv/issues/556 already found). This just creates that app instance and runs the main event loop. I've played around some time trying to eliminate the need to run the main event loop but couldn't figure it out, yet. Maybe someone with a better understanding of Cocoa can build on these findings.

**Comment by ShresthSamyak** [comment]
I've submitted a PR (#307) that finally resolves this deadlock natively, without requiring PyQt, PyObjC, or forcing the user to create their own manual Cocoa `NSApplication` event loops.

**The root cause:**
When [wait_for_playback()](cci:1://file:///c:/Users/HP/Downloads/plaika%20ai/python-mpv/mpv.py:1007:4-1011:83) calls `concurrent.futures.Future.result(timeout)` on the macOS main thread, it causes a hard thread sleep. Because Cocoa inherently requires the main thread to constantly "pump" (process) UI events, `libmpv` gets completely starved in the background and never emits the `end_file` event, resulting in an infinite deadlock.

**The fix:**
The PR intercepts the wait if it's running on the macOS main thread. Instead of hard-sleeping, it actively polls the Cocoa event loop using `CoreFoundation.CFRunLoopRunInMode` in tiny `0.05s` increments, passing control back and forth smoothly until the future resolves. 

This makes the original standard script from 2017 work perfectly out of the box:
```python
import mpv

player = mpv.MPV()
player.loadfile("test.webm")
print("waiting")
player.wait_for_playback()
print("done")

**Comment by jaseg** [comment]
Regarding the above comment by user @ShresthSamyak, that user opened up mutliple pull requests with untested AI slop (#307, #308). I can't be bothered to figure out whether there was any substance to these PRs since the user wasn't able to give meaningful responses to factual questions, and also apparently did zero testing on actual MacOS, not even in a VM.

The user is now banned from this repo, sorry for the noise y'all.

**Comment by roachcord3** [comment]
I feel responsible for the AI pests, since I offered a bounty for this issue since it's been bugging me.

To be clear to all you vibecoding nitwits: **the bounty is no longer active; I will not honor it for anyone.** You will not get any undeserved pocket money from using Claude or whatever. You frankly have no future career prospects since you've offloaded your own abilities to a machine you don't even own, and everyone considers you a menace.

**Comment by 18166714330cl-maker** [comment]
I can take an initial reproduction pass on this.

Plan: check the current macOS threading path and the historical reproduction notes, then only open a PR if I can reproduce or isolate a small compatibility fix without guessing against the old environment.
